addons/binding/org.openhab.binding.rme/src/main/java/org/openhab/binding/rme/handler/SerialThingHandler.java

```diff
@@ -0,0 +1,238 @@
+/**
+ * Copyright (c) 2014 openHAB UG (haftungsbeschraenkt) and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.openhab.binding.rme.handler;
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.io.OutputStream;
+import java.util.Enumeration;
+import java.util.TooManyListenersException;
+import gnu.io.CommPortIdentifier;
+import gnu.io.PortInUseException;
+import gnu.io.SerialPort;
+import gnu.io.SerialPortEvent;
+import gnu.io.SerialPortEventListener;
+import gnu.io.UnsupportedCommOperationException;
+import org.apache.commons.io.IOUtils;
+import org.eclipse.smarthome.core.thing.ChannelUID;
+import org.eclipse.smarthome.core.thing.Thing;
+import org.eclipse.smarthome.core.thing.ThingStatus;
+import org.eclipse.smarthome.core.thing.ThingStatusDetail;
+import org.eclipse.smarthome.core.thing.binding.BaseThingHandler;
+import org.eclipse.smarthome.core.types.Command;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The {@link SerialThingHandler} is responsible for handling commands, which
+ * are sent to one of the channels. Thing Handler classes that use serial
+ * communications can extend/implement this class, but must make sure they
+ * supplement the configuration parameters into the {@link SerialConfiguration}
+ * Configuration of the underlying Thing, if not already specified in the
+ * thing.xml definition
+ * 
+ * @author Karel Goderis - Initial contribution
+ */
+public abstract class SerialThingHandler extends BaseThingHandler implements
+		SerialPortEventListener {
+
+	// List of all Configuration parameters
+	public static final String PORT = "port";
+	public static final String BAUD_RATE = "baud";
+	public static final String BUFFER_SIZE = "buffer";
+
+	private Logger logger = LoggerFactory.getLogger(SerialThingHandler.class);
+
+	private SerialPort serialPort;
+	private CommPortIdentifier portId;
+	private InputStream inputStream;
+	private OutputStream outputStream;
+
+	public SerialThingHandler(Thing thing) {
+		super(thing);
+	}
+
+	/**
+	 * Called when data is received on the serial port
+	 * 
+	 * @param line
+	 *            - the received data as a String
+	 * 
+	 **/
+	abstract public void onDataReceived(String line);
+
+	/**
+	 * Write data to the serial port
+	 * 
+	 * @param msg
+	 *            - the received data as a String
+	 * 
+	 **/
+	public void writeString(String msg) {
+		String port = (String) this.getConfig().get(PORT);
+
+		try {
+			// write string to serial port
+			outputStream.write(msg.getBytes());
+			outputStream.flush();
+		} catch (IOException e) {
+			logger.error("Error writing '{}' to serial port {}: {}",
```

Comment:
you'll probably want to set the thing status here (rather than do error logging)